File: internal/limiter/token_bucket.go
// token_bucket.go
package limiter

import (
	"context"
	"log"
	"net/http/httputil"
	"time"

	"github.com/Sp92535/GoGate-RateLimiter/internal/utils"
	"github.com/google/uuid"
)

type TokenBucket struct {

	// track current tokens in bucket this is a unique key for redis
	key string

	// bucket capacity
	capacity int

	// bucket refill requests
	noOfRequests int

	// bucket refill duration
	interval time.Duration

	// context for closure
	ctx    context.Context
	cancel context.CancelFunc

	// corresponding proxy
	proxy *httputil.ReverseProxy
}

// constructor to initialize token bucket
func NewTokenBucket(rateLimit *utils.RateLimit, proxy *httputil.ReverseProxy) Limiter {
	ctx, cancel := context.WithCancel(context.Background())
	tb := &TokenBucket{
		key:          uuid.NewString(),
		capacity:     rateLimit.Capacity,
		ctx:          ctx,
		cancel:       cancel,
		proxy:        proxy,
		noOfRequests: rateLimit.NoOfRequests,
		interval:     rateLimit.TimeDuration,
	}

	// starting the refilling of bucket as a go routine once it is initalized
	go tb.refill()

	return tb
}

// core functionality of the algorithm the refilling of bucket
func (tb *TokenBucket) refill() {

	// initialize ticker to tick every second
	ticker := time.NewTicker(tb.interval)
	defer ticker.Stop()

	for {
		select {

		// refill as per rate
		case <-ticker.C:
			// update to whatever is minimum

			Scripts["TOKEN-BUCKET"].Run(tb.ctx, Rdb, []string{tb.key}, "core", tb.capacity, tb.noOfRequests).Int()

		// returning from function if context is cancelled
		case <-tb.ctx.Done():
			return
		}
	}

}

// function to take token and process the request
func (tb *TokenBucket) AddRequest(req *Request) bool {

	res, err := Scripts["TOKEN-BUCKET"].Run(tb.ctx, Rdb, []string{tb.key}, "take").Int()
	if err != nil {
		log.Println("Error:", err)
		return false
	}
	if res == 1 {
		go ServeReq(tb.proxy, req, nil)
		return true
	} else {
		return false
	}
}

// function to stop the algorithm
func (tb *TokenBucket) Stop() {
	tb.cancel()
}

File: internal/limiter/scripts/token_bucket.lua
-- token_bucket.lua

-- function to refill tokens in the bucket
local function refill_tokens(key, capacity, refill)
    -- getting current tokens in bucket
    local tokens = tonumber(redis.call("GET", key) or 0)

    -- refilling with new tokens
    local newTokens = math.min(capacity, tokens + refill)
    redis.call("SET", key, newTokens)

    return newTokens
end

-- function to permit request
local function take(key)
    -- getting current tokens in bucket
    local tokens = tonumber(redis.call("GET", key) or 0)

    -- take the token if bucket is not empty
    if tokens > 0 then
        redis.call("DECR", key)
        return 1
    else
        return 0
    end
end

local command = ARGV[1]
local key = KEYS[1]
if command == "take" then
    return take(key)
elseif command == "core" then
    local capacity = tonumber(ARGV[2])
    local refill = tonumber(ARGV[3])
    return refill_tokens(key, capacity, refill)
else
    return redis.error_reply("Invalid command")
end
